conans/client/cmd/uploader.py

```diff
@@ -169,23 +169,24 @@ def _collect_packages_to_upload(self, refs, confirm, remotes, all_packages, quer
                     package_revision = metadata.packages[package_id].revision
                     assert package_revision is not None, "PREV cannot be None to upload"
                     prefs.append(PackageReference(ref, package_id, package_revision))
-                refs_by_remote[ref_remote].append((ref, conanfile, prefs))
+                refs_by_remote[ref_remote].append((ref, conanfile, layout, metadata, prefs))
```

Comment:
No need to return `layout` and `metadata`, just `layout` will be enough, right?

Reply:
And about the `ref`, I see that we do `ref = ref.copy_with_rev(metadata.recipe.revision)` just after retrieving the `layout`, but maybe it makes sense to update the `layout.ref` variable with the `metadata.recipe.revision` always (inside the class), maybe add a:

```python
class PackageCacheLayout:

    @property
    def ref(self):
        return self._ref.copy_with_rev(self.metadata.recipe.revision)
```

Reply:
And if the `layout` _stores_ this `ref` and the `metadata` there are a lot of functions downstream that will be simplified.